rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,166 @@
+# [RFC] Add algorithm to dot_general's parameters in the StableHLO specification
+
+Status: Review<br/>
+Initial version: 03/15/2024<br/>
+Last updated: 03/15/2024<br/>
+Discussion thread: [GitHub](https://github.com/openxla/stablehlo/pull/2096)
+
+## Motivation
+
+We would like to allow explicitly selecting the algorithm used for individual
+`dot_general` instructions, enforcing a well-defined numeric precision. We think
+that the current precision values (`DEFAULT`, `HIGH`, `HIGHEST`) are not
+specific enough for this, because they mean different things between hardware
+platforms and even between GPU types.
+
+We also want to make the API flexible enough to support additional algorithms
+that might be faster on certain hardware
+([6xBF16](https://arxiv.org/pdf/1904.06376.pdf), 3xTF32, etc) that can't fit
+into the limited precision_config values of `DEFAULT`, `HIGH` and `HIGHEST`. For
+example "6xBF16" provides similar precision to F32, but about 2x performance on
+some GPUs.
+
+When an algorithm is not supported by an accelerator, the program should fail
+instead of implicitly falling back to another. It should be the responsibility
+of higher level frameworks such as JAX, to define the required algorithm for
+each accelerator that is used.
+
+## Proposed Specification change
+
+### dot_general
+
+#### Semantics
+
+*Add these lines:*
+
+`algorithm` defines the main properties of the algorithm used to implement the
+dot operation, which also defines the precision. If the algorithm value is set
+(`!= UNSET`), then the `precision_config` is ignored. If the algorithm value is
+not set (`= UNSET`), then the `precision_config` will take effect.
+
+In general the first two types in the algorithm name are `AType` and `BType` -
+the precisions that the LHS and RHS of the operation are rounded to, and the
+third one is `AccumType` - the accumulation type. These types are independent
+from the storage types of the inputs and the output (but not all combinations
+are supported). Some algorithms have other suffixes too.
+
+This is the current list of algorithms:
+
+* `UNSET`: There is no restriction on the algorithm, the precision is decided
+  based on the `precision_config`. Use this for every type combination that
+  doesn't have a specific algorithm.
+* `ANY_F8_ANY_F8_F32`: The inputs are used with the full precision of their
+  storage type (which must be an 8-bit floating point type - the same for both
+  operands). The accumulator type is F32.
+* `ANY_F8_ANY_F8_F32_FAST_ACCUM`: The inputs are used with the full precision of
+    their storage type (which must be an 8-bit floating point type - the same
+    for both operands). The accumulator type is formally F32, but the
+    intermediate results will not periodically be promoted to a higher
```

Comment:
>> The accumulator type is formally F32, but the intermediate results will not periodically be promoted to a higher precision. 


For a type like `ANY_F8_ANY_F8_F32`, the accumulation would happen in `F32`. Does that mean the intermediate results will always be promoted to F32. Something like, 
```
accumulation += promote_to_f32(lhs) * promote_to_f32(rhs) 
```

whereas for `ANY_F8_ANY_F8_F32`, the behavior is:

```
accumulation += lhs *rhs 
```

Also what does formally mean in `accumulator type is formally F32`? 

what is that higher precision in `...will not periodically be promoted to a higher precision`? F32?

Reply:
This was added to support CUBLASLT_MATMUL_DESC_FAST_ACCUM, and I think that Triton's
maxNumImpreciseAcc=32 setting may be similar.

> Also what does formally mean in accumulator type is formally F32?
By "formally", I mean that we pass F32 compute type to cuBLASlt.

> what is that higher precision in ...will not periodically be promoted to a higher precision? F32?
I think so. I copied this from NVIDIA's page.

The operands are not promoted as in your example.
The difference is that ANY_F8_ANY_F8_F32 is roughly equivalent to accumulating every single value into an F32 accumulator, while ANY_F8_ANY_F8_F32_FAST_ACCUM may allow to accumulate a lot of values in F16 and promote that to F32 only rarely - so it could be noticeably less precise.

If we need more info we can follow up with NVIDIA, but I think that this is perhaps enough for now.
@reedwm maybe knows better

Reply:
Thanks!

>> I think so. I got this wording from the CUBLASlt docs.

How about we make this explicit "will not periodically be promoted to a F32 precision"?

Reply:
This `ANY_F8_ANY_F8_F32_FAST_ACCUM` seems very specific to cuda.  How stable are results between different cuda versions and cards?  Is this any different than the existing `HIGH` precision option?

Reply:
IMO an "algorithm" should have well-defined semantics so that vendors can actually implement it in a consistent way.  This, on the other hand, seems like it's saying "I don't quite need F32 intermediate precision, but I want relatively high precision so give me the best you can", which to me would be the same as `HIGH`.  Do you really want the program to _fail_ if a vendor doesn't have a "Fast Accum" version and instead only had `ANY_F8_ANY_F8_F32`?

Otherwise, you may as well specify the algorithm as `CUBLASLT_MATMUL_DESC_FAST_ACCUM` if that's exactly what you wanted and want the program to fail otherwise.

Reply:
I'm not opposed to giving `ANY_F8_ANY_F8_F32_FAST_ACCUM` a more cublas-specific name. But it sounds like Triton can do something similar.

@nluehr can we get a more specific definition of what `CUBLASLT_MATMUL_DESC_FAST_ACCUM` does? The description states "intermediate results will not periodically be promoted to a higher precision". Does this mean that the flag never causes any accumulator value to be F32? Does this imply that even without this flag, values are initially accumulated in lower-than-fp32 precsision, but then "periodically" added to an fp32 accumulator?

Reply:
> But it sounds like Triton can do something similar.

"Similar" does not mean the same, so if it is different you could also have a `TRITON_MAX_NUM_IMPRECISE_ACC_32`, though again that sounds like it's less of a guaranteed precision/algorithm, and more like the existing `HIGH` precision config we already have.

Reply:
In the new struct based proposal this would map to can_accumulate_in_lower_precision_for_some_steps=true.
(feel free to reopen if needed)